### PR TITLE
lowering: extend temporary lifetimes around await

### LINKED ISSUE #63832: temporary lifetime around await is (maybe) unexpectedly short
So @sfackler pointed out something to me. The problem @sfackler reported was that temporary lifetimes around `await` were too short. The effect was that `foo(&[123]).await` would give an error where `foo(&[123])` would not (in sync code). My assumption was that this was caused by our temporary lifetimes being derived from the desugared version of await, which doesn't seem like what one might expect.

Here is an example showing the problem:

[async version, playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=4fd818313ecfd99400df97db7f976b65)

[sync version, playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7babe7be6ac54aff1d9e60d98f884153)

### PR DESCRIPTION
**Pull request by davidtwco**
Fixes #63832.

r? @matthewjasper 
cc @nikomatsakis

### CODE REVIEW

File: src/test/ui/async-await/issue-63832-await-short-temporary-lifetime-1.rs [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    async fn borrow_async(&self) {}
+
+    fn with(&mut self, s: &str) -> &mut Self {
+        self.0 = s.into();
+        self
+    }
+}
+
+async fn test() {
+    // error[E0716]: temporary value dropped while borrowed
```
Comment (Centril) [inline src/test/ui/async-await/issue-63832-await-short-temporary-lifetime-1.rs]: Error *and* check-pass?

**Review by Centril — COMMENTED** [review]
Could you please also update the doc comment on `fn lower_expr_await` to accurately reflect the new desugaring and then follow up with a PR against the reference (See https://github.com/rust-lang-nursery/reference/pull/635 for @nikomatsakis's initial PR.)?

File: src/test/ui/async-await/issue-63832-await-short-temporary-lifetime-1.rs (hunk 1 above)
Comment (davidtwco) [reply src/test/ui/async-await/issue-63832-await-short-temporary-lifetime-1.rs]: Oops, fixed.

**Comment by davidtwco** [comment]
Thanks @Centril, fixed those comments and opened rust-lang-nursery/reference#674.

**Review by Centril — APPROVED** [review]

File: src/librustc/hir/lowering/expr.rs [hunk 2]
```diff
@@ -658,8 +651,24 @@ impl LoweringContext<'_> {
...
+            let expr = P(self.lower_expr(expr));
+            P(self.expr_match(
+                span,
+                expr,
+                hir_vec![pinned_arm],
+                hir::MatchSource::AwaitDesugar,
+            ))
+        };
+
         hir::ExprKind::Block(
```
Comment (matthewjasper) [inline src/librustc/hir/lowering/expr.rs]: I don't think that there's a need to wrap the match expression in a block now.

Comment (davidtwco) [reply src/librustc/hir/lowering/expr.rs]: Fixed, thanks!

Comment (matthewjasper) [reply src/librustc/hir/lowering/expr.rs]: The comment for the function also needs updating

File: src/librustc/hir/lowering/expr.rs [hunk 3]
```diff
@@ -508,15 +508,16 @@ impl LoweringContext<'_> {
...
     /// ```rust
     /// {
-    ///     let mut pinned = <expr>;
-    ///     loop {
-    ///         match ::std::future::poll_with_tls_context(unsafe {
-    ///             ::std::pin::Pin::new_unchecked(&mut pinned)
-    ///         }) {
-    ///             ::std::task::Poll::Ready(result) => break result,
-    ///             ::std::task::Poll::Pending => {},
+    ///     match future {
```
Comment (davidgauch) [inline src/librustc/hir/lowering/expr.rs]: For the sake of clarity the `future` on this line should probably be `<expr>`

File: src/librustc/hir/lowering/expr.rs (hunk 2 above)
Comment (davidtwco) [reply src/librustc/hir/lowering/expr.rs]: Oops, fixed.

File: src/librustc/hir/lowering/expr.rs (hunk 3 above)
Comment (davidtwco) [reply src/librustc/hir/lowering/expr.rs]: Fixed, thanks!

**Comment by RalfJung** [comment]
Is it expected that this is a breaking change? A Miri testcase is failing now:
```
2019-09-10T19:17:04.0564321Z normalized stderr:
2019-09-10T19:17:04.0564553Z error[E0597]: `x` does not live long enough
2019-09-10T19:17:04.0565934Z   --> $DIR/async-fn.rs:18:13
2019-09-10T19:17:04.0566063Z    |
2019-09-10T19:17:04.0566130Z 18 |     async { x + y }.await
2019-09-10T19:17:04.0566458Z    |     --------^------
2019-09-10T19:17:04.0566539Z    |     |     | |
2019-09-10T19:17:04.0566641Z    |     |     | borrowed value does not live long enough
2019-09-10T19:17:04.0566727Z    |     |     value captured here by generator
2019-09-10T19:17:04.0566873Z    |     a temporary with access to the borrow is created here ...
2019-09-10T19:17:04.0566971Z 19 | }
2019-09-10T19:17:04.0567254Z    | -
2019-09-10T19:17:04.0567323Z    | |
2019-09-10T19:17:04.0567409Z    | `x` dropped here while still borrowed
2019-09-10T19:17:04.0567524Z    | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `impl std::future::Future`
2019-09-10T19:17:04.0567653Z    |
2019-09-10T19:17:04.0568341Z    = note: The temporary is part of an expression at the end of a block. Consider forcing this temporary to be dropped sooner, before the block's local variables are dropped. For example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block.
2019-09-10T19:17:04.0568497Z 
2019-09-10T19:17:04.0568622Z error[E0597]: `y` does not live long enough
2019-09-10T19:17:04.0568922Z   --> $DIR/async-fn.rs:18:17
2019-09-10T19:17:04.0569038Z    |
2019-09-10T19:17:04.0569103Z 18 |     async { x + y }.await
2019-09-10T19:17:04.0569731Z    |     ------------^--
2019-09-10T19:17:04.0569811Z    |     |     |     |
2019-09-10T19:17:04.0569908Z    |     |     |     borrowed value does not live long enough
2019-09-10T19:17:04.0569999Z    |     |     value captured here by generator
2019-09-10T19:17:04.0570108Z    |     a temporary with access to the borrow is created here ...
2019-09-10T19:17:04.0570189Z 19 | }
2019-09-10T19:17:04.0570464Z    | -
2019-09-10T19:17:04.0570534Z    | |
2019-09-10T19:17:04.0570619Z    | `y` dropped here while still borrowed
2019-09-10T19:17:04.0570726Z    | ... and the borrow might be used here, when that temporary is dropped and runs the destructor for type `impl std::future::Future`
2019-09-10T19:17:04.0570849Z    |
2019-09-10T19:17:04.0571489Z    = note: The temporary is part of an expression at the end of a block. Consider forcing this temporary to be dropped sooner, before the block's local variables are dropped. For example, you could save the expression's value in a new local variable `x` and then make `x` be the expression at the end of the block.
2019-09-10T19:17:04.0571685Z 
2019-09-10T19:17:04.0571753Z error: aborting due to 2 previous errors
```

**Comment by yoshuawuyts** [comment]
Not sure if this was intended to cause breakage, but we had to patch our code in async-std likely because of this PR: https://github.com/async-rs/async-std/pull/182/files.

__before__
```rust
 peer.send(format!("from {}: {}\n", from, msg)).await?
```
__after__
```rust
let msg = format!("from {}: {}\n", from, msg);
peer.send(msg).await?
```

**Comment by RalfJung** [comment]
We needed a [similar patch](https://github.com/rust-lang/miri/pull/946) in Miri.

**Comment by seanmonstar** [comment]
So are the above examples a regression?

**Comment by Centril** [comment]
Depends... can you create a minimal reproducer for before/after?

**Comment by tmandry** [comment]
This causes errors to crop up all over the Fuchsia code base. Based on https://github.com/rust-lang/rust/issues/63832#issuecomment-524662434, it seems like this might have been expected, however.

**Comment by seanmonstar** [comment]
@Centril The change in miri is quite succinct, and the "before" doesn't *appear* to be bad... while the "after" seems odd? https://github.com/rust-lang/miri/pull/946/files#diff-ecc42d2008d431ab52ee698995abd41b

Is something more needed?

**Pull request merged**